snuba/clickhouse/optimize/optimize.py

```diff
@@ -286,48 +294,58 @@ def optimize_partition_runner(
     1. There are no more partitions which need optimization.
     2. The final cutoff time is reached. In this case, the scheduler will
     raise an exception which would be propagated to the caller.
+
+    Details of execution flow:
+    1. start by reading configured_num_threads from the Snuba Admin runtime config
+    2. dispatches configured_num_threads threads to optimize configured_num_threads partitions (1 partition per thread)
+    3. as soon as one thread finishes, check configured_num_threads from runtime config again
+    4. if configured_num_threads > number of currently active threads, dispatch more threads
     """
-    remaining_partitions = partitions
-    while remaining_partitions:
-        schedule = scheduler.get_next_schedule(remaining_partitions)
-        num_threads = len(schedule.partitions)
-        logger.info(
-            f"Running schedule with cutoff time: "
-            f"{schedule.cutoff_time} with {num_threads} threads"
-        )
-        threads: MutableSequence[threading.Thread] = []
-        for i in range(0, num_threads):
-            threads.append(
-                threading.Thread(
-                    target=optimize_partitions,
-                    args=(
+
+    with ThreadPoolExecutor(max_workers=32) as executor:
+        futures: set[Future[Any]] = set()
+        # jitter start time
+        for jitter in scheduler.get_start_time_jitter_for_each_partition():
```

Comment:
@mcannizz 

Previously, I had jitter scheduling after determining the number of threads. This is because the number of jitter futures should be the same as the number of threads. However, the code's readability suffered. 

This current approach of scheduling jittering before determining the number of threads has the problem of scheduling either 1 or 2 jitter futures, since in `get_start_time_jitter_for_each_partition`, `self.__parallel` is either 1 or 2.

I think a quick and easy fix is to just make `get_start_time_jitter_for_each_partition` work with the number of threads instead of `self.__parallel`, so the function is more like `get_start_time_jitter_for_each_thread`.

Reply:
I'm not sure I understand the problem. Can you explain it a bit more? It would be helpful to understand what behavior we expect and what actual behavior we get with this code.